code/over_time.py

```diff
@@ -0,0 +1,152 @@
+    return output
+
+
+def git_current_branch(dp: PathLike) -> str:
+    """Determines the name of the currently checked-out branch."""
+    status = git_status(dp)
+    return status.split("\n")[0].replace("On branch ", "")
+
+
+def git_checkout(dp: PathLike, commit_or_branch: str):
+    """Check out a specific branch or commit in the repository under `dp`."""
+    output = subprocess.check_output(["git", "-C", str(dp), "checkout", commit_or_branch])
+    output = output.strip().decode("ascii")
+    return output
+
+
+
+def eval_by_commit(
+    dp: PathLike,
+    func: Callable[[PathLike], T],
+    commits: Sequence[str],
+    *,
+    raise_on_error: bool = True,
+) -> Dict[str, Optional[T]]:
+    """Apply `func` to the `dp` for each of the `commits` and return the results.
+    
+    Requires the repository at `dp` to be in a clean `git status` state.
+    In the end, the current branch will be checked out again.
+
+    Parameters
+    ----------
+    dp
+        Path to a local git repository.
+    func
+        A callable to apply at each commit.
+        It should take one parameter `dp` and return something.
+    commits
+        A sequence of commits to execute the function at.
+    raise_on_error
+        If ``True``, exceptions other than SyntaxErrors are raised.
+
+    Returns
+    -------
+    results
+        Maps commit IDs to return values of the provided callable,
+        or ``None`` in case of syntax errors at the respective commit.
+    """
+    status = git_status(dp)
+    if "working tree clean" not in status and "nothing added to commit but untracked" not in status:
+        raise Exception(f"The git status of '{dp}' is unclean:\n\n{status}")
+    branch = git_current_branch(dp)
+    results = {}
+    for commit in progress_bar(commits):
+        try:
+            git_checkout(dp, commit)
+            results[commit] = func(dp)
+        except SyntaxError:
+            results[commit] = None
+        except:
+            print(f"Failed to apply function at commit {commit}")
+            if raise_on_error:
+                raise
+    print(f"\nEvaluations completed. Checking out branch '{branch}'.")
+    git_checkout(dp, branch)
+    return results
+
+
+def complexity_by_commit(dp: PathLike, commits: Optional[Sequence[str]] = None) -> pd.DataFrame:
+    """Convenience wrapper around ``eval_by_commit`` to determine mean code complexity metrics over time.
+
+    Parameters
+    ----------
+    dp
+        Path to a local git repository.
+    commits
+        Optional sequence of commits IDs to run for.
+        Defaults to the entire git history of the repo.
+
+    Returns
+    -------
+    df
+        DataFrame indexed by `commit_id`,
+        with `timespan` and mean `COMPLEXITY_METRICS` columns.
+    """
+    if not commits:
+        commits = git_log(dp)
+
+    results = eval_by_commit(
+        dp=dp,
+        func=lambda dp: get_repo_complexities(dp).set_index("repo")[COMPLEXITY_METRICS].mean(),
```

Comment:
The `get_repo_complexities()` function failed for me on the first initial commit or so as that one had a basically empty python file.

I managed to fix it by changing this line inside `get_repo_complexities()` 
```
df = pd.DataFrame(complexities)
```
to 
```
df = pd.DataFrame(complexities, columns=COMPLEXITY_COLUMNS)
```